source/MaterialXCore/Element.h

```diff
@@ -247,6 +247,42 @@ class Element : public std::enable_shared_from_this<Element>
         return EMPTY_STRING;
     }
 
+    /// @}
+    /// @name Units
+    /// @{
+
+    /// Set the unit.
+    void setUnit(const string& unit)
+    {
+        setAttribute(UNIT_ATTRIBUTE, unit);
+    }
+
+    /// Return true if a unit attribute exists.
+    bool hasUnit() const
+    {
+        return hasAttribute(UNIT_ATTRIBUTE);
+    }
+
+    /// Return the unit .
+    const string& getUnit() const
+    {
+        return getAttribute(UNIT_ATTRIBUTE);
+    }
+
+    /// Return the unit string that is active at the scope of this
+    /// element, taking all ancestor elements into account.
+    const string& getActiveUnit() const
+    {
+        for (ConstElementPtr elem = getSelf(); elem; elem = elem->getParent())
```

Comment:
Again, we need to figure out how this will work for multiple unit types. An input can only have a single unit, for example _length_, but at parent/document level there can be multiple unit types. So inheriting from parent like this might not work later.

Reply:
Yes. This needs to be resolved. But as said it is actually not correct for "colorspace" as well. Having any non-input element having a unit is possible, but as with colorspace not allowed AFAIK.

Reply:
Then I think we should remove units from the Element base class and move it over to the ValueElement class.

Reply:
I need to find the closest base class for input/param/bindinput/bindparam. I originally had this on ValueElement which should cover this so need to move it back. Also getActiveUnit() will not exist as there will be no inheritance.